resolwe/flow/managers/listener.py

```diff
@@ -47,153 +67,235 @@
-                    self._redis = await self._make_connection()
-                return await meth(self._redis, *args, **kwargs)
-            except aioredis.RedisError:
-                logger.exception("Redis connection error")
-                if self._redis:
-                    self._redis.close()
-                    await self._redis.wait_closed()
-                    self._redis = None
-                await asyncio.sleep(3)
-
-    async def clear_queue(self):
-        """Reset the executor queue channel to an empty state."""
-        conn = await self._make_connection()
-        try:
-            script = """
-                local keys = redis.call('KEYS', ARGV[1])
-                redis.call('DEL', unpack(keys))
-            """
-            await conn.eval(
-                script,
-                keys=[],
-                args=["*{}*".format(settings.FLOW_MANAGER["REDIS_PREFIX"])],
-            )
-        finally:
-            conn.close()
+        This method uses Django ORM.
+        """
+        if self._worker is None:
+            self._worker = Worker.objects.get_or_create(
+                data=self.data, defaults={"status": Worker.STATUS_RESPONDING}
+            )[1]
+        return self._worker
 
-    async def __aenter__(self):
-        """On entering a context, start the listener thread."""
-        self._should_stop = False
-        self._redis = await self._make_connection()
-        self._runner_coro = asyncio.ensure_future(self.run())
-        return self
+    def _choose_worst_status(self, status1: str, status2: str) -> str:
+        """Get more problematic status of the two.
 
-    async def __aexit__(self, typ, value, trace):
-        """On exiting a context, kill the listener and wait for it.
+        Assumption: statuses in Data.STATUS_CHOICES are ordered from least
+            to most problematic.
 
-        .. note::
+        :raises AssertionError: when status1 or status2 do not represent
+            possible status of Data object.
 
-            Exceptions are all propagated.
         """
-        await self._runner_coro
-        self._redis.close()
-        await self._redis.wait_closed()
-        # Make sure the connection is cleaned up.
-        self._redis = None
+        statuses = [e[0] for e in Data.STATUS_CHOICES]
+        assert status1 in statuses
+        assert status2 in statuses
+        status_index = max(statuses.index(status1), statuses.index(status2))
+        return statuses[status_index]
+
+    def _save_error(self, error: str):
+        """Log error to Data object, ignore possible exceptions."""
+        with suppress(Exception):
+            self.data.refresh_from_db()
+        self.data.process_error.append(error)
+        self.data.status = Data.STATUS_ERROR
+        with suppress(Exception):
+            self.data.save(update_fields=["process_error", "status"])
+
+    def _log_exception(
+        self, error: str, extra: dict = {}, save_to_data_object: bool = True
```

Comment:
I am aware of that. In this case the only thing I an doing is updating `extra` with data id, which is done every time.
I will change it though since the code can change in the future and this can evolve into a hard-to-spot bug.